File: client/typescript/src/_impl/mir/metadata/dispatch.ts
import type {
    DispatchMetadata,
    DispatchSlot,
    DynamicShape,
    DynamicTable,
    Vtable,
} from "../../../_generated/mir/metadata/dispatch.js";
import type { LocalNodeId } from "../../../_generated/mir/tree/node.js";
import { getLocalNodeValue, localNodeIdMatches } from "../node.js";

export const DispatchSlotImpl = {
    /** Return the zero-based slot index. */
    index(slot: DispatchSlot): number {
        return slot;
    },
};

export const DispatchMetadataImpl = {
    /** Return the vtable for a type when present. */
    vtable(metadata: DispatchMetadata, ty: LocalNodeId): Vtable | undefined {
        return metadata.vtables.find((table) => localNodeIdMatches(table.ty, ty));
    },

    /** Return the dynamic table for a concrete type and constraint. */
    dynamicTable(
        metadata: DispatchMetadata,
        concrete: LocalNodeId,
        constraint: LocalNodeId,
    ): DynamicTable | undefined {
        return metadata.dynamicTables.find((table) => {
            const isConcrete = localNodeIdMatches(table.concrete, concrete);
            const isConstraint = localNodeIdMatches(table.constraint, constraint);

            return isConcrete && isConstraint;
        });
    },

    /** Return dynamic shape metadata for a constraint type id. */
    dynamicShape(metadata: DispatchMetadata, constraint: LocalNodeId): DynamicShape | undefined {
        return getLocalNodeValue(metadata.dynamicShapes, constraint);
    },
};

File: client/python/src/destack/_impl/mir/metadata/dispatch.py
from __future__ import annotations

from destack._impl.model import ModelImpl


class DispatchMetadataImpl(ModelImpl):
    """Methods for dispatch metadata."""

    def vtable(self, ty):
        """Return the vtable for a type when present."""

        for table in self.vtables:
            # compare structural ids
            if table.ty == ty:
                return table

        return None

    def dynamic_table(self, concrete, constraint):
        """Return the dynamic table for a concrete type and constraint."""

        for table in self.dynamic_tables:
            is_concrete = table.concrete == concrete
            is_constraint = table.constraint == constraint

            # return the matching table
            if is_concrete and is_constraint:
                return table

        return None

    def dynamic_shape(self, constraint):
        """Return dynamic shape metadata for a constraint type id."""

        return self.dynamic_shapes.get(constraint)
